README
Introduction
------------

Hosttag is a client for tagging hostnames into groups or classes, storing 
the mappings in a redis datastore. It requires the 'redis' rubygem.


Hosttag assumes your redis server is found on a server called 'hosttag' on
the default redis port (6379). You can either setup 'hosttag' as a hostname
alias in a local hosts file, or specify an alternative host to use in a
HOSTTAG_SERVER environment variable, or explicitly via a -s/--server
<hostname> argument. Similarly, an alternate port can be specified using a
HOSTTAG_PORT environment variable, or a -p/--port <port> command line
argument.



Basic Usage
-----------

  # Set some tags (-T says assume anything you don't know is a tag)
  $ htset -T zeus dell centos centos5 server
  $ htset -T minerva dell centos centos6 laptop
  $ htset -T hermes hp ubuntu lucid desktop
  $ htset -T athena hp ubuntu maverick server


  # Queries
  $ ht dell               
  minerva zeus

  $ ht ubuntu
  hermes athena

  $ ht -o hp dell         # -o/--or ORs arguments
  hermes minerva zeus

  $ ht -a ubuntu server   # -a/--and ANDs arguments
  athena

  $ ht -o de*             # tags and hosts may include globs (*, ?, [M-N])
  zeus minerva hermes

  $ ht -A                 # all hosts
  athena hermes minerva zeus

  $ ht -T                 # all tags
  centos centos5 centos6 dell desktop hp laptop lucid maverick server




Security
--------

A warning on security: basically, there isn't any. Anyone who can get write
to your redis server can write, rewrite, and delete all your hosttags. So
you should only use hosttag on a trusted network.

Redis supports a password-ed mode, and it might make sense to use that via
a config file or something, so you could at least restrict usage via file
permissions (although traffic would still be in the clear, I believe). For
now, though, there's no security.



Installation
------------

To install from source, unpack the tarball and run 'rake install' (requires
rubygems and rake). You can set GEM_HOME to an alternate install directory 
(e.g. ~/.gem) if you want to install somewhere other than your gem default.

RPMs are available for CentOS/RHEL {5,6} from the Open Fusion repository:

  http://www.openfusion.net/linux/openfusion_rpm_repository




Author
------

Gavin Carr <gavin@openfusion.com.au>, http://www.openfusion.net/

Copyright 2010-2011 Gavin Carr.



Licence
-------

This software is free software, licensed under the GNU Public License v3
or any later version.



ChangeLog
ChangeLog
---------

0.13 - 2012-05-11
* Add glob support on tags and hosts (Nathan Norton).

0.12 - 2011-11-21
* Fix smembers buglet for 1.9.x compatibility (Alexis Letessier).
* Change from ftools to fileutils for compatibility with 1.9.x (Alexis Letessier).
* Change hashbag line to /usr/bin/env ruby (Alexis Letessier).
* Add back require rubygems to hosttag/server to workaround problems on 1.8.x.
* Add missing Hosttag include to test/test_hosttag_bin.rb.

0.11 - 2011-02-03
* Update README.
* Add Hosttag::Server support for HOSTTAG_{SERVER,PORT,NAMESPACE} env variables.
* Add gemspec and assorted Rakefile bits.
* Rename hosttag-server subpackage to hosttag-server-utils.
* Remove rubygems require from Hosttag::Server.

0.10.5 - 2011-01-13
* Remove htdump from spec file.

0.10.4 - 2011-01-12
* Make --list work properly with 'ht -A' and 'ht -T'.

0.10.3 - 2011-01-12
* Tweak hosttag.render to handle nil resultsets better.

0.10.2 - 2011-01-11
* Fix a few small bugs in htset.

0.10.1 - 2011-01-07
* Tweak Hosttag.lookup_keys to re-raise original error on second lookup failure.
* Fix stupid :all == 1 buglets in htset.

0.10 - 2011-01-06
* Rename all_hosts* and all_tags* labels for better clarity.
* Add hosttag_truncate method, and use for import.
* Replace htimport in unit tests with Hosttag.hosttag_import_from_directory.
* Remove bogus chdir from Hosttag.load_directory.
* Update Hosttag.hosttag_lookup_{tags,hosts} to leave options hash unchanged.
* Rename Hosttag.lookup to lookup_keys.
* Refactor core import functionality from htimport to Hosttag.hosttag_import_from_directory.
* Unit test additions and updates.
* Hosttag documentation updates.



test/ts_all.rb
require 'test_hosttag_lib.rb'
require 'test_hosttag_bin.rb'
require 'test_htset_lib.rb'
require 'test_htset_bin.rb'


test/test_ht.rb
#!/usr/bin/ruby

require 'test/unit'
require 'ftools'

class TestHosttag < Test::Unit::TestCase

  # format: args (string) => expected (either string or regex)
  TESTS = [
    [ 'centos',                 'a m n' ],
    [ 'laptop',                 'n' ],
    [ 'vps',                    'm' ],
    [ 'laptop vps',             '' ],
    [ '-a laptop vps',          '' ],
    [ '-o laptop vps',          'm n' ],
    [ '-t n',                   'centos centos5 centos5-i386 laptop' ],
    [ '-t a',                   'centos centos5 centos5-x86_64 public' ],
    [ '-t m',                   'centos centos4 centos4-x86_64 public vps' ],
    [ '-t a n',                 'centos centos5 centos5-i386 centos5-x86_64 laptop public' ],
    [ '-to a n',                'centos centos5 centos5-i386 centos5-x86_64 laptop public' ],
    [ '-t -o a n',              'centos centos5 centos5-i386 centos5-x86_64 laptop public' ],
    [ '-ta a n',                'centos centos5' ],
    [ '-t -a a n',              'centos centos5' ],
    [ '-t -a a n m',            'centos' ],
    # All cases
    [ '-A',                     'a m n' ],
    [ '-A -A',                  'a g h m n' ],
    [ '-T',                     'centos centos4 centos4-x86_64 centos5 centos5-i386 centos5-x86_64 laptop public vps' ],
    # Try unrecognised tags as hosts
    [ 'n',                      'centos centos5 centos5-i386 laptop' ],
    [ 'a n',                    'centos centos5 centos5-i386 centos5-x86_64 laptop public' ],
    # Invalid data
    [ 'foo',                    %r{ \btag\b  .* not\sfound }x ],
    [ 'centos foo',             %r{ \btag\b  .* not\sfound }x ],
    [ '-t foo',                 %r{ \bhost\b .* not\sfound }x ],
    [ '-t n foo',               %r{ \bhost\b .* not\sfound }x ],
  ]

  def test_hosttag
    test_args = '--server localhost --ns hosttag_testing'

    # Setup
    datadir = "#{File.dirname(__FILE__)}/data_hosttag"
    bindir = File.join(File.dirname(__FILE__), '..', 'bin')
    File.directory?(datadir) or throw "missing datadir #{datadir}"
    `#{bindir}/htimport #{test_args} --delete --datadir #{datadir}`
    `#{bindir}/htdump #{test_args}`

    TESTS.each do |args, expected|
      got = `#{bindir}/hosttag #{test_args} #{args} 2>&1`.chomp
      if expected.is_a?(String)
        assert_equal(expected, got, "args: #{args}")
      elsif expected.is_a?(Regexp)
        assert_match(expected, got, "args: #{args}")
      end
    end
  end

end



test/test_htset.rb
#!/usr/bin/ruby

require 'test/unit'
require 'ftools'

require 'hosttag'
include Hosttag

class TestHtset < Test::Unit::TestCase

  TESTS = [
    [ 
      [ 'htset', 'foo', 'centos' ],    # operation
      [ 'foo' ],                       # ht -A output
      [ 'foo' ],                       # ht -A -A output
      [ 'centos' ],                    # ht -T output
    ],
    [ 
      [ 'htset', 'foo', 'centos5' ],
      [ 'foo' ],
      [ 'foo' ],
      [ 'centos', 'centos5' ],
    ],
    [ 
      [ 'htset', 'bar', 'centos' ],
      [ 'bar', 'foo' ],
      [ 'bar', 'foo' ],
      [ 'centos', 'centos5' ],
    ],
    [ 
      [ 'htset', 'bar', 'centos4' ],
      [ 'bar', 'foo' ],
      [ 'bar', 'foo' ],
      [ 'centos', 'centos4', 'centos5' ],
    ],
    [ 
      [ 'htset', 'bar', 'SKIP' ],
      [ 'foo' ],
      [ 'bar', 'foo' ],
      [ 'centos', 'centos4', 'centos5' ],
    ],
    [ 
      [ 'htdel', 'bar', 'SKIP' ],
      [ 'bar', 'foo' ],
      [ 'bar', 'foo' ],
      [ 'centos', 'centos4', 'centos5' ],
    ],
    [ 
      [ 'htdel', 'bar', 'centos4' ],
      [ 'bar', 'foo' ],
      [ 'bar', 'foo' ],
      [ 'centos', 'centos5' ],
    ],
    [ 
      [ 'htdel', 'bar', 'centos' ],
      [ 'foo' ],
      [ 'foo' ],
      [ 'centos', 'centos5' ],
    ],
    [ 
      [ 'htdel', 'foo', 'centos' ],
      [ 'foo' ],
      [ 'foo' ],
      [ 'centos5' ],
    ],
    [ 
      [ 'htdel', 'foo', 'centos5' ],
      [],
      [],
      [],
    ],
  ]

  def setup
    @test_args = '--server localhost --ns hosttag_testing'
    @test_opts = { :server => 'localhost', :namespace => 'hosttag_testing', :debug => false }
    datadir = "#{File.dirname(__FILE__)}/data_reset"
    @bindir = File.join(File.dirname(__FILE__), '..', 'bin')
    File.directory?(datadir) or throw "missing datadir #{datadir}"
    `#{@bindir}/htimport #{@test_args} --delete --datadir #{datadir}`
  end

  def test_htset
    TESTS.each do |op, hosts1, hosts2, tags1|
      cmd = op.join(' ')
      bin = op.shift()
      if bin == 'htset'
        hosttag_add_tags([ op[0] ], [ op[1] ], @test_opts)
      else
        hosttag_delete_tags([ op[0] ], [ op[1] ], @test_opts)
      end

      # Check results
      assert_equal(hosts1, hosttag_all_hosts(@test_opts), "all_hosts, #{cmd}")
      assert_equal(hosts2, hosttag_all_hosts(@test_opts.merge({ :include_skip? => true })), 
        "all_hosts include_skip, #{cmd}")
      assert_equal(tags1, hosttag_all_tags(@test_opts),  "all_tags, #{cmd}")
    end
  end
end



lib/hosttag.rb

require 'hosttag/server'
require 'set'
require 'pp'

module Hosttag

  # Lookup the given tag(s), returning an array of hosts to which they apply.
  # If multiple tags are given, by default the list of hosts is those to
  # which ALL of the tags apply i.e. results are ANDed or intersected. To
  # change this, pass :rel => :or in the options hash.
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :rel - either :and or :or, specifying the relationship to use when
  #   interpreting the set of tags. :rel => :and returns the set of hosts to
  #   which ALL the given tags apply; :rel => :or returns the set of hosts
  #   to which ANY of the tags apply. Default: :rel => :and.
  # - :include_skip? - flag indicating whether to include hosts that have
  #   the SKIP tag set. Default: false i.e. omit hosts tagged with SKIP.
  def hosttag_lookup_tags(*args)
    options = args.last.is_a?(Hash) ? args.pop : {}
    return lookup_keys(args, options.merge({ :type => :tag }))
  end

  # Lookup the given host(s), returning an array of tags that apply to
  # them. If multiple hosts are given, by default the list of tags is
  # those applying to ANY of the given hosts i.e. the results are ORed or
  # unioned. To change this pass an explicit :rel => :and in the options
  # hash.
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :rel - either :and or :or, specifying the relationship to use when
  #   interpreting the set of hosts. :rel => :and returns the set of tags
  #   that apply to ALL the given hosts; :rel => :or returns the set of tags
  #   that apply to ANY of the given hosts. Default: :rel => :or.
  # - :include_skip? - flag indicating whether to include hosts that have
  #   the SKIP tag set. Default: false i.e. omit hosts tagged with SKIP.
  def hosttag_lookup_hosts(*args)
    options = args.last.is_a?(Hash) ? args.pop : {}
    return lookup_keys(args, options.merge({ :type => :host }))
  end

  # Lookup the given host(s) or tag(s), returning an array of tags or hosts,
  # as appropriate. If a :type option is not explicitly given, first tries
  # the lookup using hosttag_lookup_tags, and if that fails retries using
  # hosttag_lookup_hosts.
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :type - either :host or :tag, specifying how to interpret the given
  #   arguments: :type => :host specifies that the arguments are hosts, and
  #   that the resultset should be a list of tags; :type => :tag specifies
  #   that the arguments are tags, and the resultset should be a list of
  #   hosts. Required, no default.
  # - :rel - either :and or :or, specifying the relationship to use when
  #   interpreting the set of results. :rel => :and returns only results
  #   that have ALL of the given attributes i.e. the AND result set;
  #   :rel => :or returns results that have ANY of the given attributes
  #   i.e. the OR result set. Default: depends on :type - :and for :type
  #   => :host, and :or for :type => :tag.
  # - :include_skip? - flag indicating whether to include hosts that have
  #   the SKIP tag set. Default: false i.e. omit hosts tagged with SKIP.
  def hosttag_lookup(*args)
    options = args.last.is_a?(Hash) ? args.pop : {}
    return lookup_keys(args, options) if options[:type]

    begin
      return hosttag_lookup_tags(args, options)
    rescue => e
      begin
        return hosttag_lookup_hosts(args, options)
      rescue
        # If both lookups failed, re-raise original error
        raise e
      end
    end
  end

  # Return an array of all hosts
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :include_skip? - flag indicating whether to include hosts that have
  #   the SKIP tag set. Default: false i.e. omit hosts tagged with SKIP.
  def hosttag_all_hosts(options)
    r = hosttag_server(options)
    key = r.get_key(options[:include_skip?] ? 'all_hosts_full' : 'all_hosts')
    $stderr.puts "+ key: #{key}" if options[:debug]
    ret =  r.smembers(key).sort
    return ret.map { |r| r.gsub(/_meta/,"") }.to_set.to_a.sort
  end

  # Return an array of all tags
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :include_skip? - flag indicating whether to include the SKIP tag.
  #   Default: false. Included for completeness.
  def hosttag_all_tags(options)
    r = hosttag_server(options)
    key = r.get_key(options[:include_skip?] ? 'all_tags_full' : 'all_tags')
    $stderr.puts "+ key: #{key}" if options[:debug]
    ret = r.smembers(key).sort
    return remove_metadata_tags( ret, options )
  end

  # Add the given tags to all the given hosts
  def hosttag_add_tags(hosts, tags, options)
    r = hosttag_server(options)

    tags_meta = extract_namespaces_from_tags( tags ) 

    # Add tags to each host
    skip_host = {}
    all_hosts_skip_hosts = true
    hosts.each do |host|
      host_meta = host + "_meta"

      key = r.get_key('host', host)
      tags.each { |tag| r.sadd(key, tag) }
      key_meta = r.get_key('host', host_meta)
      tags_meta.each { |tag| r.sadd(key_meta, tag) }

      if r.sismember(key, 'SKIP')
        skip_host[host] = true
      else
        all_hosts_skip_hosts = false
      end

      # Add to all_hosts sets
      all_hosts = r.get_key('all_hosts')
      all_hosts_full = r.get_key('all_hosts_full')
      # all_hosts shouldn't include SKIP hosts, so those we remove
      if skip_host[host]
        r.srem(all_hosts, host)
        r.srem(all_hosts, host_meta)
      else
        r.sadd(all_hosts, host)
        r.sadd(all_hosts, host_meta)
      end
      r.sadd(all_hosts_full, host)
      r.sadd(all_hosts_full, host_meta)
    end

    # Add hosts to each tag
    recheck_for_skip = false
    tags_total = tags + tags_meta
    tags_total.each do |tag|
      # If we've added a SKIP tag to these hosts, flag to do some extra work
      recheck_for_skip = true if tag == 'SKIP'

      key = r.get_key('tag', tag)
      hosts.each do |host|
        host_meta = host + "_meta"
        # The standard case is to add the host to the list for this tag.
        # But we don't want SKIP hosts being included in these lists, so
        # for them we actually do a remove to make sure they're omitted.
        if skip_host[host] and tag != 'SKIP'
          r.srem(key, host)
        else
          r.sadd(key, host) if tags.include? tag
          r.sadd(key, host_meta) if tags_meta.include? tag
        end
      end

      # Add to all_tags sets
      all_tags = r.get_key('all_tags')
      all_tags_full = r.get_key('all_tags_full')
      r.sadd(all_tags, tag) unless all_hosts_skip_hosts
      r.sadd(all_tags_full, tag)
    end

    # If we've added a SKIP tag here, we need to recheck all tags for all skip hosts
    recheck_skip_change_for_all_tags(skip_host.keys, :add, r) if recheck_for_skip
  end

  # Delete the given tags from all the given hosts
  def hosttag_delete_tags(hosts, tags, options)
    r = hosttag_server(options)

    tags_meta = extract_namespaces_from_tags( tags )

    # Delete tags from each host
    non_skip_host = {}
    hosts.each do |host|
      host_meta = host + "_meta"

      key = r.get_key('host', host)
      tags.each { |tag| r.srem(key, tag) }
      key_meta = r.get_key('host', host_meta)
      tags_meta.each { |tag| r.srem(key_meta, tag) if tag_safe_to_delete?( host , tag, options ) }
      #tags_meta.each { |tag| r.srem(key_meta, tag) }

      if r.sismember(key, 'SKIP')
        skip_host = true
      else
        non_skip_host[host] = true
      end

      # Delete from all_hosts sets
      all_hosts = r.get_key('all_hosts')
      all_hosts_full = r.get_key('all_hosts_full')
      # If all tags have been deleted, or this is a SKIP host, remove from all_hosts
      if r.scard(key) == 0 or skip_host
        r.srem(all_hosts, host)
        r.srem(all_hosts, host_meta)
      else
        # NB: we explicitly add here in case we've deleted a SKIP tag
        r.sadd(all_hosts, host)
        r.sadd(all_hosts, host_meta)
      end
      if r.scard(key) == 0
        r.srem(all_hosts_full, host)
        r.del(key)
        r.srem(all_hosts_full, host_meta)
        r.del(key_meta)
      end
    end

    # Delete hosts from each tag
    recheck_for_skip = false
    all_tags = r.get_key('all_tags')
    all_tags_full = r.get_key('all_tags_full')
    tags_total = tags_meta + tags
    tags_total.each do |tag|
      # If we've deleted a SKIP tag from these hosts, flag to do some extra work
      recheck_for_skip = true if tag == 'SKIP'

      tag_key = r.get_key('tag', tag)
      hosts.each { |host| r.srem(tag_key, host + "_meta") if tag_safe_to_delete?( host , tag, options ) }
      #hosts.each { |host| r.srem(tag_key, host + "_meta") }
      hosts.each { |host| r.srem(tag_key, host) }

      # Delete from all_tags sets
      # If all hosts have been deleted (or this is the SKIP tag), remove from all_tags
      if r.scard(tag_key) == 0 or tag == 'SKIP'
        r.srem(all_tags, tag)
      else
        # NB: we explicitly add here in case we've deleted a SKIP tag
        r.sadd(all_tags, tag)
      end
      if r.scard(tag_key) == 0
        r.srem(all_tags_full, tag)
        r.del(tag_key)
      end
    end
    r.del(all_tags) if r.scard(all_tags) == 0
    r.del(all_tags_full) if r.scard(all_tags_full) == 0

    # If we've deleted a SKIP tag here, we need to recheck all tags for all non-skip hosts
    recheck_skip_change_for_all_tags(non_skip_host.keys, :delete, r) if recheck_for_skip
  end

  # Delete all hosts and tags in the hosttag datastore. This is the nuclear option,
  # used in hosttag_import_from_directory if :delete => true. Interactively confirms
  # unless the :autoconfirm option is set.
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :autoconfirm - if true, truncate without asking for any confirmation
  def hosttag_truncate(options)
    if not options[:autoconfirm]
      print "Do you really want to delete EVERYTHING from your datastore? [yN] "
      $stdout.flush
      confirm = $stdin.gets.chomp
      return unless confirm =~ %r{^y}i
    end

    r = hosttag_server(options)
    r.keys(r.get_key("*")).each { |k| r.del(k) }
  end

  # Delete all tags from the given hosts. Interactively confirms the deletions
  # unless the :autoconfirm option is set.
  # The final argument may be an options hash, which accepts the following
  # keys:
  # - :autoconfirm - if true, do deletes without asking for any confirmation
  def hosttag_delete_all_tags(hosts, options)
    if not options[:autoconfirm]
      host_str = hosts.join(' ')
      print "Do you want to delete all tags on the following host(s):\n  #{host_str}\nConfirm? [yN] "
      $stdout.flush
      confirm = $stdin.gets.chomp
      return unless confirm =~ %r{^y}i
    end

    hosts.each do |host|
      begin
        tags = hosttag_lookup_hosts(host, options)
        hosttag_delete_tags([ host ], tags, options)
      rescue
        warn "Warning: invalid host '#{host}' - cannot delete"
      end
    end
  end

  # Import hosts and tags from the given directory. The directory is
  # expected to contain a set of directories, representing hosts; each
  # file within those directories is treated as a tag that applies to
  # that host.
  # Options is a hash which accepts the following keys:
  # - :delete - if true, delete ALL hosts and tags from the datastore
  #   before doing the import.
  # - :autoconfirm - if true, don't interactively confirm deletions
  def hosttag_import_from_directory(datadir, options)
    # Delete ALL hosts and tags from the datastore if options[:delete] set
    hosttag_truncate(options) if options[:delete]

    # Load directory into a { host => [ taglist ] } hash
    host_tag_hash = load_directory(datadir, options)

    # Add all hosts and tags
    host_tag_hash.each do |host, tags|
      hosttag_add_tags([ host ], tags, options)
    end
  end

  private

  # Lookup the given keys in the redis datastore, returning an array of
  # results. If more than one key is specified, resultsets are merged
  # (either ANDed or ORed) depending on the value of the :rel option.
  # The final argument must be an options hash, which accepts the
  # following options:
  # - :type - specifies the type of keys to lookup, either :host or :tag.
  #   Required.
  # - :rel - specifies how to merge multiple resultsets, either :and (set
  #   intersection) or :or (set union).
  def lookup_keys(*args)
    options = args.last.is_a?(Hash) ? args.pop : {}
    args.flatten!

    type = options[:type]
    throw "Required option 'type' missing" if not type
    rel = options[:rel]

    r = hosttag_server(options)

    # Default a rel if we have multiple args
    if args.length > 1 and not rel
      rel = (type == :tag ? :and : :or)
    end
    $stderr.puts "+ rel (#{type}): #{rel}" if rel and options[:debug]

    # Map keys to fetch
    keys = args.collect {|v| r.get_key(type, v) }
    $stderr.puts "+ keys: #{keys.join(' ')}" if options[:debug]

    # Check all keys exist
    keys.each do |k|
      if not r.exists(k)
        item = k.sub(%r{^[^:]+::[^:]+:}, '')
        raise "Error: #{type} '#{item}' not found."
      end
    end

    # Lookup and return
    ret = ""
    if keys.length == 1
      ret = r.smembers(keys[0]).to_set.to_a.sort
    elsif rel == :and
      ret = r.sinter(*keys).to_set.to_a.sort
    else
      ret = r.sunion(*keys).to_set.to_a.sort
    end
    
    ret.map { |r| r.gsub(/_meta/,"") } 
  end

  # for a given set of tags extract out a list of all the 
  # namespaces for use in the metadata host data set.
  # Eg: tags=["centos","test::sydney::dc1::rack2","prod::sydney::dc2::rack4"]
  # we will return an array consisting of the namespaces.
  # EG: res=["test::sydney::dc1","test::sydney","test::","sydney::dc1::rack2",
  # "sydney::dc1","sydney::","rack2::","prod::sydney::dc2","prod::sydney",
  # "sydney::dc2::rack4","sydney::dc2","sydney::","dc2::rack4","rack4::"]
  # These tags will not show up in normal results BUT you can search on them 
  # to get lists of hosts, for example ht sydney::dc2 will return a list of 
  # all hosts in sydney and dc2.
  def extract_namespaces_from_tags( tags )
    result = []
    tags.each do |t|
      next unless  t.include? "::"

      a = t.split("::")
      
      until a.empty?  do
        first = a.shift
        c = Array.new a

        result << first + "::"
        c.each do |cc|
          element = result.last + "::" + cc
          element.gsub!(/::::/, '::')
          result << element unless tags.include? element
        end
      end
    end
    result.to_set.to_a # return unique elements (.uniq() does not work here)
  end

  # remove the metadata results from the set of all tags that is returned 
  # when the hosttag_all_tags is used. 
  def remove_metadata_tags( tags , options) 
    results = []
    tags.each do |t|
      unless t.include? "::"
        results << t 
        next 
      end
      res = hosttag_lookup_tags(t , options)
      res.each do |r|
        if hosttag_lookup_hosts(r , options).include? t
          results << t
        end
      end 
    end
    results.to_set.to_a.sort
  end

  # when we delete a tag we also need to delete the correspoding 
  # meta tags, however we need to check that a meta tag is still not 
  # being used before deleting, for example TAG aaa::bbb will have 
  # the following meta tags aaa:: bbb:: however if the host also has 
  # aaa::ccc  (with tags aaa:: ccc::) we will remove needed meta tags
  # this method will stop this from happening
  def tag_safe_to_delete?( host, tag, options) 
    r = hosttag_server(options)
    k = r.get_key(:host, host)
    return true unless r.exists(k) # if host does not exist any more safe to delete!

    count = 0
    tags = hosttag_lookup_hosts(host , options)
    tags.each do |t|
      ns_tags = extract_namespaces_from_tags( t )  
      count += 1 if ns_tags.include? tag 
    end
    return false if count > 0 
    return true
  end

  # If we've added or removed a SKIP tag, we now have to recheck all tags for
  # the given hosts, removing or re-adding them from/to those tag sets, and
  # then recalculate the all_tags set for each of those tags
  def recheck_skip_change_for_all_tags(hosts, change, r)
    recheck_tags = {}
    hosts.each do |host|
      host_key = r.get_key('host', host)
      r.smembers(host_key).each do |tag|
        next if tag == 'SKIP'

        tag_key = r.get_key('tag', tag)
        # If we've added SKIP tags, then we remove the host from tagsets
        # (or vice-versa)
        if change == :add
          r.srem(tag_key, host)
        else
          r.sadd(tag_key, host)
        end

        recheck_tags[tag] = tag_key
      end
    end

    # Now recheck the all_tags set, adding tags that have hosts, and
    # removing any that don't
    all_tags = r.get_key('all_tags')
    recheck_tags.each do |tag, tag_key|
      tag_host_count = r.scard(tag_key)
      if tag_host_count == 0
        r.srem(all_tags, tag)
      else
        r.sadd(all_tags, tag)
      end
    end
    r.del(all_tags) if r.scard(all_tags) == 0
  end

  # Load all host/tag files in datadir, returning a { host => [ taglist ] } hash
  def load_directory(datadir, options)
    host_tag_hash = {}

    Dir.glob("#{datadir}/*").each do |host_path|
      next if not File.directory?(host_path)
      host = host_path.sub(/^#{datadir}\//, '')

      host_tag_hash[host] = []
      Dir.glob("#{host_path}/*").each do |tag_path|
        next if not File.file?(tag_path)
        tag = File.basename(tag_path)
        host_tag_hash[host].push(tag)
      end
    end

    return host_tag_hash
  end

  def die(error)
    warn error
    exit 1
  end

  def hosttag_server(options)
    begin
      r = Hosttag::Server.new(options)
    rescue Resolv::ResolvError => e
      die e
    end
  end

end


data/load_data.pl
#!/usr/bin/perl

use strict;

use File::Basename;
use TokyoCabinet;

my $DB = '/var/lib/misc/hosttag.tch';

my $data_dir = dirname($0);
die "Cannot find data dir '$data_dir'" unless -d $data_dir;

my %tag = ();

# Load all data
for my $host_dir (glob("$data_dir/*")) {
  next if $host_dir eq '.' || $host_dir eq '..';
  next if ! -d $host_dir;

  my $host = basename $host_dir;
  
  for my $tag_file (glob("$host_dir/*")) {
    $tag{ $host } ||= [];
    push @{ $tag{ $host } }, basename( $tag_file );
  }
}

# Invert %tag to get %host hash
my %host = ();
for my $host (keys %tag) {
  for my $tag ( @{ $tag{$host} } ) {
    $host{ $tag } ||= [];
    push @{ $host{ $tag } }, $host;
  }
}

# Write database
my $db = TokyoCabinet::HDB->new;
$db->open( $DB, $db->OWRITER | $db->OCREAT )
  or die "DB $DB open failed: " . $db->errmsg( $db->ecode );

for my $host (keys %tag) {
  $db->put("hosts/$host", join(' ', sort @{ $tag{$host} }) . "\n")
    or die "DB $DB put failed: " . $db->errmsg( $db->ecode );
}
for my $tag (keys %host) {
  $db->put("tags/$tag", join(' ', sort @{ $host{$tag} }) . "\n")
    or die "DB $DB put failed: " . $db->errmsg( $db->ecode );
}
$db->put("hosts/ALL", join(' ', sort keys %tag));

# Check for orphaned hosts/tags
$db->iterinit;
while (my $key = $db->iternext) {
  if ($key =~ m! ^hosts/(.*)$ !x) {
    my $host = $1;
    next if exists $tag{$host};
    next if $host eq 'ALL';
    print "Deleting orphaned host '$host'\n";
    $db->out($key);
  }
  elsif ($key =~ m! ^tags/(.*)$ !x) {
    my $tag = $1;
    next if exists $host{$tag};
    print "Deleting orphaned tag '$tag'\n";
    $db->out($key);
  }
}

$db->close
  or die "DB $DB close failed: " . $db->errmsg( $db->ecode );



lib/hosttag/server.rb
# Hosttag::Server class, extending Redis

# Blech - shouldn't need to require rubygems here, but fails on 1.8.x otherwise
require 'rubygems'
require 'redis'
require 'resolv'

module Hosttag
  class Server < Redis
    def initialize(options)
      @defaults = { 
        :server     => ENV['HOSTTAG_SERVER']    || 'hosttag', 
        :port       => ENV['HOSTTAG_PORT']      || 6379, 
        :namespace  => ENV['HOSTTAG_NAMESPACE'] || 'hosttag',
      }
      @defaults.merge!(options)
       
      # Check :server name resolves
      begin
        Resolv.getaddress( @defaults[:server] )
      rescue Resolv::ResolvError
        raise Resolv::ResolvError, 
          "Host '#{@defaults[:server]}' does not resolve (try --server <hostname>?)", caller
      end

      # Connect to redis
      super( :host => @defaults[:server], :port => @defaults[:port] )
    end

    def get_key(*elt)
      "#{@defaults[:namespace]}::#{elt.join(':')}"
    end
  end
end

